src/ir_passes/bundle_elim.rs

```diff
@@ -0,0 +1,233 @@
+    context: DenseIndexInfo<Component, HashMap<PortIdx, Vec<PortIdx>>>,
+}
+
+impl BundleElim {
+    fn port(&self, pidx: PortIdx, comp: &mut Component) -> Vec<PortIdx> {
+        let one = comp.add(Expr::Concrete(1));
+        let Port {
+            owner,
+            width,
+            live,
+            info,
+        } = comp.get(pidx).clone();
+
+        let Liveness { idx, len, range } = live;
+
+        let start = comp.get(range.start).clone();
+        let end = comp.get(range.end).clone();
+
+        let len = len.as_concrete(comp).unwrap();
+
+        // if we need to preserve external interface information, we can't have bundle ports in the signature.
+        if comp.src_info.is_some()
+            && matches!(owner, crate::ir::PortOwner::Sig { .. })
+        {
+            assert!(
+                len == 1,
+                "Bundle ports in the signature are not supported."
+            );
+
+            // need to preserve the original portidx here to save the source information.
+            return vec![pidx];
+        }
+
+        let ports = (0..len)
+            .map(|i| {
+                let binding: HashMap<_, _> = [(idx, i)].into();
+
+                let offset = start.offset.resolve(comp, &binding);
+                let start = comp.add(Time {
+                    event: start.event,
+                    offset,
+                });
+
+                let offset = end.offset.resolve(comp, &binding);
+                let end = comp.add(Time {
+                    event: end.event,
+                    offset,
+                });
+                let live = Liveness {
+                    idx,
+                    len: one,
+                    range: Range { start, end },
+                };
+
+                let owner = match &owner {
+                    crate::ir::PortOwner::Sig { dir } => {
+                        crate::ir::PortOwner::Sig { dir: dir.clone() }
+                    }
+                    crate::ir::PortOwner::Inv { inv, dir, base } => {
+                        crate::ir::PortOwner::Inv {
+                            inv: *inv,
+                            dir: dir.clone(),
+                            base: Foreign::new(
+                                self.context
+                                    .get(base.owner)
+                                    .get(&base.key)
+                                    .unwrap()[i as usize],
+                                base.owner,
+                            ),
+                        }
+                    }
+
+                    crate::ir::PortOwner::Local => crate::ir::PortOwner::Local,
+                };
+
+                comp.add(Port {
+                    live,
+                    owner,
+                    info,
+                    width,
+                })
+            })
+            .collect();
+        // delete the original port
+        comp.delete(pidx);
+        // delete the corresponding parameter
+        comp.delete(idx);
```

Comment:
Looks good! We should also change the `Validate` pass so that it checks that invalid ports etc. are not being used by anyone, either inside the component or through a `Foriegn`.